File: src/UI/OnScreenKeyboard.ts
import { frequencies, getPlayTone } from '../Instruments/Oscillator';
import { Commands, octaveAndNoteToMessage } from '../Tools/Midi';

export type IOnClickKey = ({
  note,
  octave,
}: {
  note: keyof typeof frequencies;
  octave: number;
}) => void;

export function listen(playTone: ReturnType<typeof getPlayTone>): IOnClickKey {
  return ({ note, octave }) => {
    const message = octaveAndNoteToMessage({ note, octave });
    const command = Commands.NOTE_ON;
    const { stopTone } = playTone({ command, message, value: 100 });
    setTimeout(() => stopTone(), 400);
  };
}

File: web/UI/OnScreenKeyboard.dart
import 'dart:html';
import 'dart:async';

import '../Tools/Midi.dart';

void render(HandleMidi handleMidi) {
  var i = 0;
  noteStrings.forEach((String note) {
    i += 1;
    final key = DivElement();
    key.className = 'key';
    querySelector('.keyboard')?.children.add(key);
    key.onClick.listen((event) {
      final message =
          octaveAndNoteToMessage(note: note, octave: (i / 12).floor() + 4);
      final stopTone =
          handleMidi(command: Commands.NOTE_ON, message: message, value: 127);

      // Handle stopping the midi 400 ms later
      Timer(Duration(milliseconds: 400), () {
        stopTone();
      });
    });
  });
}
